runtime/memprof.c

```diff
@@ -42,45 +550,612 @@ struct memprof_thread_s {
+        if (runner != NULL) {
+          CAMLassert(runner->running_table == es);
+          CAMLassert(runner->running_index == i);
+          runner->running_index = j;
+        }
+      }
+      ++ j;
+    }
+    ++ i;
+    if (es->young == i) es->young = j;
+    if (es->next == i) es->next = j;
+  }
+  es->evict = es->live = j;
+  CAMLassert(es->next <= es->live);
+  CAMLassert(es->young <= es->live);
+
+  ensure_entries(es, 0);
+}
+
+/* Transfer all entries from one entries table to another. */
+
+static void entries_transfer(entries_t from, entries_t to)
+{
+  size_t offset = to->live;
+
+  if (from->live == 0) {
+    return;
+  }
+
+  CAMLassert(to->config == from->config);
+
+  ensure_entries(to, from->live);
+  to->live += from->live;
+
+  for (size_t j = 0; j < from->live; ++j) {
+    to->t[offset + j] = from->t[j];
+    memprof_thread_t thread = from->t[j].running;
+    if (thread) { /* unusual */
+      CAMLassert(thread->running_table == from);
+      CAMLassert(thread->running_index == j);
+      thread->running_table = to;
+      thread->running_index = offset+j;
+    }
+  }
+
+  if (to->young == offset) {
+    to->young = offset + from->young;
+  }
+  if (to->evict == offset) {
+    to->evict = offset + from->evict;
+  }
+  if (to->next == offset) {
+    to->next = offset + from->next;
+  }
+  /* Reset `from` to empty, and allow it to shrink */
+  ensure_entries(from, -from->live);
+  from->young = from->evict = from->next = from->live = 0;
+}
+
+/* If es->config points to a DISCARDED configuration, update
+ * es->config to CONFIG_NONE. Return es->config. */
+
+static value validated_config(entries_t es)
+{
+  value config = es->config;
+  if ((config != CONFIG_NONE) &&
+      (Status(config) == CONFIG_STATUS_DISCARDED)) {
+    es->config = config = CONFIG_NONE;
+  }
+  return config;
+}
+
+/* Return configuration for a thread. If it's been discarded, then
+ * reset it to CONFIG_NONE and return that.
+*/
+
+static value thread_config(memprof_thread_t thread)
+{
+  return validated_config(&thread->entries);
+}
+
+/*** Create and destroy orphan tables ***/
+
+/* Orphan any surviving entries from a domain or its threads (after
+ * first discarding any deleted and offset entries), onto the domain's
+ * orphans list. This copies the domain's table itself, to avoid
+ * copying the potentially live array. */
+
+static void orphans_create(memprof_domain_t domain)
+{
+  /* Transfer any entries from the domain's threads to the domain. */
+  memprof_thread_t thread = domain->threads;
+  while (thread) {
+    entries_transfer(&thread->entries, &domain->entries);
+    thread = thread->next;
+  }
+
+  entries_t es = &domain->entries;
+  entries_evict(es, true); /* remove deleted and offset entries */
+  if (!es->live) { /* no live entries */
+    entries_clear(es);
+  } else { /* Orphan surviving entries */
+    memprof_orphan_table_t ot =
+      caml_stat_alloc(sizeof(memprof_orphan_table_s));
+    if (ot) {
```

Comment:
These out-of-memory questions are important; thanks for picking them up. I have refactored here so that we allocate the orphans entries table before doing any transfers, and various functions now return `bool`. `caml_memprof_start` raising `Out_of_memory` I think is correct. I think we have to live with `domain_destroy` possibly silently running out of memory. We won't leak any statmemprof data structures, but obviously will fail to call callbacks on the lost entries.

Reply:
(in the specific case of (a) destroying a domain which (b) has active tracking entries and (c) allocating the orphan entries table fails).